File: BOJ/17246.cpp
#include <bits/stdc++.h>
using namespace std;

#define INF 987654321

int main() {
    int n, m;
    scanf("%d %d", &n, &m);

    int g[200][200];
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            g[i][j] = i == j ? 0 : INF;
        }
    }

    for (int i = 0; i < m; ++i) {
        int a, b, k;
        scanf("%d %d %d", &a, &b, &k);
        g[a - 1][b - 1] = g[b - 1][a - 1] = k;
    }

    for (int k = 0; k < n; ++k) {
        for (int i = 0; i < n; ++i) {
            if (i == k || g[i][k] == INF) continue;
            for (int j = 0; j < n; ++j) {
                if (i == j || j == k || g[k][j] == INF) continue;
                g[i][j] = min(g[i][j], g[i][k] + g[k][j]);
            }
        }
    }

    int q;
    scanf("%d", &q);
    while (q--) {
        int x, y;
        scanf("%d %d", &x, &y);
        int ans = g[x - 1][y - 1] < INF ? g[x - 1][y - 1] : -1;
        printf("%d\n", ans);
    }

    return 0;
}

File: BOJ/17246.py
from math import inf as INF

n, m = map(int, input().split())
g = [[INF if i != j else 0 for j in range(n+1)] for i in range(n+1)]
for _ in range(m):
    a, b, k = map(int, input().split())
    g[a][b] = g[b][a] = k

for k in range(1, n+1):
    for i in range(1, n+1):
        if g[i][k] == INF or i == k:
            continue
        for j in range(1, n+1):
            if g[k][j] == INF or i == j or j == k:
                continue
            g[i][j] = min(g[i][j], g[i][k] + g[k][j])

q = int(input())
for _ in range(q):
    x, y = map(int, input().split())
    ans = -1 if g[x][y] is INF else g[x][y]
    print(ans)